src/query/plan/preprocess.cpp

```diff
@@ -546,6 +548,13 @@ std::vector<SingleQueryPart> CollectSingleQueryParts(SymbolTable &symbol_table,
       if (auto *merge = utils::Downcast<query::Merge>(clause)) {
         query_part->merge_matching.emplace_back(Matching{});
         AddMatching({merge->pattern_}, nullptr, symbol_table, storage, query_part->merge_matching.back());
+      } else if (auto *foreach = utils::Downcast<query::Foreach>(clause)) {
+        for (auto *clause : foreach->clauses_) {
+          if (auto *merge = utils::Downcast<query::Merge>(clause)) {
+            query_part->merge_matching.emplace_back(Matching{});
+            AddMatching({merge->pattern_}, nullptr, symbol_table, storage, query_part->merge_matching.back());
```

Comment:
Based on the comments of `merge_matching` I think this is not good, because you are modifying `merge_matching` without adding anything to `remaining_clauses`.

To be honest I don't see the whole picture, but I think we should check whether we have any similar clause that defines its scope and do something similar here.